### ISSUE TITLE
Mismatched array parameter bounds

### ISSUE THREAD

**Issue by tchatow** [randombit/botan#3065]
Compiling with Clang 15 and `-Warray-parameter`,

```
external/botan/src/lib/block/serpent/serpent_avx2/serpent_avx2.cpp:54:44: error: argument 'in' of type 'const uint8_t[128]' (aka 'const unsigned char[128]') with mismatched bound [-Werror,-Warray-parameter]
void Serpent::avx2_encrypt_8(const uint8_t in[128], uint8_t out[128]) const
                                           ^
bazel-out/k8-opt/bin/external/botan/botan/serpent.h:42:41: note: previously declared as 'const uint8_t[64]' (aka 'const unsigned char[64]') here
      void avx2_encrypt_8(const uint8_t in[64], uint8_t out[64]) const;
```

This seems to be a genuine programming error, even though the bounds don't actually do anything for an array argument.

Also appears for `newhope_shareda(uint8_t sharedkey[],…`
and `void bigint_monty_redc_…`

**Comment by randombit** [comment]
Thanks for reporting this! We should enable this warning in maintainer flags for Clang builds.

**Comment by randombit** [comment]
I fixed the AVX2 one, and the NEWHOPE is is no longer relevant because we are removing NEWHOPE completely. Can you provide the error message related to the `bigint_monty_redc` variants? There the declarations and definitions seem to be matching, so I'm not sure what needs to change.

**Comment by tchatow** [comment]
Looks like `bigint_monty_redc` was fixed in 578a969f619abd63bfe57b80be1d38ab3be88b49.

**Comment by randombit** [comment]
Closing now as fixed. Thanks for reporting these. Unfortunately in CI we do not have access to a recent enough clang, so can't enable these warnings in CI builds at this time.
